### ISSUE TITLE
Design the data model

### ISSUE THREAD

**Issue by seanhagen** [grammachachi/adventurelookup#3]
Leaving aside the question of actual database technology, we need to figure out what the data model looks like.

Now, some of this stuff is going to depend on other conversations happening first -- is there going to be some kind of moderation system, how do adventures get updated, how do new adventures get entered, how complex is the search going to be, etc, etc.

Some things to figure out:

* what does an adventure contain?
* what kinds of metadata is linked to an adventure ( ie, items, villians, authors, etc ). Do the various types of metadata need different models, or can they all use the same model with a type field? 
* if there is going to be an audit trail or some sort of moderation queue, how will that work?

**Comment by seanhagen** [comment]
I've got some ideas that I'm going to sketch up on this -- I think the easiest way is to create a drawing in Google Drive, and share that with folks who want to contribute. We can post images into this issue so that we've got a record of ideas and how that drawing changed.

**Comment by thewanderer41** [comment]
I made an information model that will (hopefully) eventually turn into a data model in the shared [Google Doc](https://docs.google.com/document/d/1cD4Nf-l_Y5eCN7uJ0Nu8D-XRx0woXNyl4aMCPSSJmz4/edit?usp=sharing).  I didn't include *items* because I felt like they were usually too generic, but maybe they should be included

**Comment by Nymall** [comment]
There has to be some kind of moderation que. So many sites are lit on fire by lack of any moderation - all it takes is one idiot to ruin everyone's day. Perhaps a account age requirement before moderation is removed?

**Comment by Ezmar** [comment]
I think one item to consider would be what can be taken care of as simple tags vs what needs to be linked to other information types.  So while Items and Enemies and Dungeons would be something an adventure has, things like Setting and Style, or other more subjective qualities wouldn't really need to be robustly interconnected with the data structure.

I think we need to define several properties of PCs, since that's a huge categorization for adventures.  We need to know Party size, level, possibly class, etc.  There's also the weird things we run into when trying to make a database that's searchable on as many conceivable parameters as possible, such as the nature of the villain.  It's one thing to determine if the BBEG is a Lich, but quite another to differentiate between adventures where the Antagonist is a powerful creature vs a cunning normal person.

These are just some thoughts, I'll need to wait a little bit and watch some stuff actually get done before I remember how databases are supposed to be set up.  :P

**Comment by Nymall** [comment]
A word of warning from my present job...

We have a massive database that was designed for Realtors to access property listings from across the country. I love the thing - it's powerful, and you can get fine grained results off of almost any data - however it was designed by a programmer first and foremost.

There are over three hundred customizable search parameters, with several restricted fields varying from price it zone to number of times it has appeared on our service. There is only one field the realtors use, however - Search by MLS#.

In designing the database, it's best to keep in mind the average user won't be a power user or a librarian. Having that functionality is fine, just don't overwhelm the user in options, or the user is more likely to mentally check out and not even bother.

**Comment by Ezmar** [comment]
Agreed.  It would probably be a good idea to take a poll of DMs somehow to figure out what most of them would go for first when looking for an adventure, with the other filter options available as... filters.  To narrow down what they're looking for if they get more than they want, in order to get something more specific.  I would imagine that one of the most important criteria would be the level, so people would start poking around the level 7-10 adventures until they found something that stuck out to them.

Still, part of the overall design philosophy is for people to be able to get specific results for whatever they might be looking for, so while it's important not to bloat the system with data points, we want to be able to sort adventures by several different traits.  A user-defined tag system could be helpful for some of the smaller things, but that's a whole host of problems on its own.  We should hit the important stuff with the data model, and fit the rest in where it makes sense to do so.

**Comment by seanhagen** [comment]
Total off the cuff, not numbers backed, guestimate on what I think are some of the most important search parameters:

* party size & recommended level
* bbeg type ( ie, lich, dragon, god, whatever )
* setting

There are plenty of other things, but they can be hidden in an "advanced search" or behind a power user query language ( think like boolean search operators for google and whatnot )

**Comment by cpnielsen** [comment]
I agree with @seanhagen, I really think we can solve a lot of search criteria with a few fields. The ones mentioned are a good start, maybe BBEG should be swapped out or complimented with a "general monster type" using MM types.

In addition to that, we can solve a lot of issues with a simple key-word based / free-text search that basically indexes the title, description and more of the adventure to allow free-text search (so you can search for "goblins" or "castle" or "dwarves").
